cpp/src/arrow/compute/exec/options.h

```diff
@@ -430,6 +430,33 @@ class ARROW_EXPORT SelectKSinkNodeOptions : public SinkNodeOptions {
   SelectKOptions select_k_options;
 };
 
+/// \brief Make a node which selects a range of rows passed through it
+///
+/// All batches pushed to this node will be accumulated, then selected, by the given
+/// fields. Then sorted batches will be forwarded to the generator in sorted order and
+/// select the rows with an offset and a limit.
+class ARROW_EXPORT FetchSinkNodeOptions : public SinkNodeOptions {
+ public:
+  explicit FetchSinkNodeOptions(
+      int64_t offset, int64_t count,
+      std::function<Future<util::optional<ExecBatch>>()>* generator,
+      SortOptions sort_options, bool sort_first = false)
+      : SinkNodeOptions(generator),
+        offset(offset),
+        count(count),
+        sort_options(sort_options),
+        sort_first(sort_first) {}
+
+  /// The number of rows to skip.
+  int64_t offset;
+  /// The number of rows to select.
+  int64_t count;
+  /// Sort options
+  SortOptions sort_options;
+  /// Determine sort or fetch precedence.
+  bool sort_first;
```

Comment:
I was thinking about this some more.  I don't think it will ever make sense to fetch first.

In the long term (when we have better support for ordered execution), fetch and sort will be separate nodes, and we should reject any plan that has a fetch node when there is no current ordering (e.g. immediately after a hash-join).

In the short term I don't see any benefit to adding support for fetch->sort since there is never a valid incoming ordering in our current exec plan.

Reply:
I see your point. I will update the code.